presto-native-execution/presto_cpp/main/TaskManager.cpp

```
@@ -639,7 +639,14 @@ std::unique_ptr<TaskInfo> TaskManager::createOrUpdateTaskImpl(
       if (source.noMoreSplits) {
         LOG(INFO) << "No more splits for " << taskId << " for node "
                   << source.planNodeId;
-        execTask->noMoreSplits(source.planNodeId);
+        // If the task has not been started yet, we collect the plan node to
+        // call 'no more splits' after the start.
+        if (prestoTask->taskStarted) {
+          execTask->noMoreSplits(source.planNodeId);
+        } else {
+          prestoTask->planNodesWithDelayedNoMoreSplits_.emplace_back(
+              source.planNodeId);
+        }
```

The behavior was changed to not call velox::Task::noMoreSplits() when prestoTask is not started. Probably better to not change this behavior by always calling noMoreSplits.

---

Yes, by this change I pursued the "Ensure calling 'no more splits' after the task started".
"After" is the important part here.

The bug itself (description can be seen in this PR) is about that.

---

> The delayed no more split signals will be processed no matter we have enabled server pushback or not? Thanks!

When pushback is not enabled or it is enabled, but there is no overload, then the task is started right away and the 'no more splits' signal will be processed right away as well (not getting into the 'set').

So, there would be no delayed 'no more splits' signals in that case.

---

> @spershin I see. So these task queue methods were deeply integrated with task manager and still be called even server pushback is disabled. Then maybe add a test to test the original behavior, making sure the delayed no more splits call is received by velox task. I think it will be nice (haven't gone through the entire pushback code so not sure if it makes sense. Ignore if it doesn't) to extract the pushback logic out to a separate module if possible (not going through the code path if disabled) as a standalone feature.

@tanjialiang 

TaskManager manages the task, so yes, it owns the task queue and the logic of task queuing/dequeuing.

When pushback is not enabled we never put or take anything to/from the queue.

The existing tests cover the tests for no-pushback mode.

If you have something specific in mind - we can add such test, however, I so far fail to understand what is missing and how would it relate to this PR and the bug we are trying to fix.